pkg/graphite/convert/whisperconverter/whisper.go

```diff
@@ -54,53 +54,72 @@ func ReadPoints(w Archive, name string) ([]whisper.Point, error) {
 		return nil, fmt.Errorf("whisper file contains no archives for metric: %q", name)
 	}
 
-	totalPoints := 0
-	for _, a := range w.GetArchives() {
-		totalPoints += int(a.Points)
-	}
-
-	// Preallocate space for all allPoints in one slice.
-	allPoints := make([]pointWithPrecision, totalPoints)
-	pIdx := 0
 	// Dump one precision level at a time and write into the output slice.
+	// Its important to remember that the archive with index 0 (first archive)
+	// has the raw data and the highest precision https://graphite.readthedocs.io/en/latest/whisper.html#archives-retention-and-precision
+	seenTs := map[uint32]struct{}{}
+	lastMinTs := uint32(0)
+	var keptPoints []whisper.Point
 	for i, a := range w.GetArchives() {
 		archivePoints, err := w.DumpArchive(i)
 		if err != nil {
 			return nil, fmt.Errorf("failed to dump archive %d from whisper metric %s", i, name)
 		}
-		for j, p := range archivePoints {
-			allPoints[pIdx+j] = pointWithPrecision{p, a.SecondsPerPoint}
+
+		var minArchiveTs, maxArchiveTs uint32
+		// calculate maxArchiveTs
+		for _, p := range archivePoints {
+			// We want to track the max timestamp of the archive because we know
+			// it virtually represents now() and we wont have newer points.
+			// Then the min timestamp of the archive would be maxTs - the archive
+			// retention.
+			if p.Timestamp > maxArchiveTs {
+				maxArchiveTs = p.Timestamp
+			}
 		}
```

Comment:
Why do we do this for each archive again and again?

Don't all archives have the same "now" and they only differ in their retention period?

Reply:
Ah, you're right.  We had seen the "offset" value in the database dump but that is a byte offset, not a timestamp offset.